python/cuml/test/test_array.py

```diff
@@ -122,39 +112,69 @@ def test_array_init(input_type, dtype, shape, order):
 
         assert np.array_equal(truth, data)
     else:
-        found_owner = False
+        helper_test_ownership(ary, inp, False)
 
-        def get_owner(curr):
-            if (isinstance(curr, CumlArray)):
-                return curr._owner
-            elif (isinstance(curr, cp.ndarray)):
-                return curr.data.mem._owner
-            else:
-                return None
 
-        # Make sure the input array is in the ownership chain
-        curr_owner = ary
+@pytest.mark.parametrize('input_type', test_input_types)
+def test_ownership_with_gc(input_type):
+    # garbage collection slows down the test suite significantly, we only
+    # need to test for each input type, not for shapes/dtypes/etc.
+    if input_type == 'numpy':
+        pytest.skip("test not valid for numpy input")
+
+    inp, ary, ptr = create_ary_init_tests(input_type, np.float32, (10, 10),
+                                          'F')
+
+    helper_test_ownership(ary, inp, True)
+
+
+def create_ary_init_tests(ary_type, dtype, shape, order):
+    if ary_type is not None:
+        inp = create_input(ary_type, dtype, shape, order)
+        ary = CumlArray(data=inp)
+        ptr = ary.ptr
+    else:
+        inp = create_input('cupy', dtype, shape, order)
+        ptr = inp.__cuda_array_interface__['data'][0]
+        ary = CumlArray(data=ptr, owner=inp, dtype=inp.dtype, shape=inp.shape,
+                        order=order)
+
+    return (inp, ary, ptr)
 
-        while (curr_owner is not None):
-            if (curr_owner is inp):
-                found_owner = True
-                break
 
-            curr_owner = get_owner(curr_owner)
+def get_owner(curr):
+    if (isinstance(curr, CumlArray)):
+        return curr._owner
+    elif (isinstance(curr, cp.ndarray)):
+        return curr.data.mem._owner
+    else:
+        return None
 
-        assert found_owner, "GPU input arrays must be in the owner chain"
 
-        inp_copy = deepcopy(cp.asarray(inp))
+def helper_test_ownership(ary, inp, garbage_collect):
```

Comment:
what is inp? input pointer?

Reply:
the non cumlarray object that is used to initialized the `ary` cuml array